# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: ipc/glue/MessageLink.h
# Commit: 7b7819250f79
# Full Hash: 7b7819250f7965227783233af6de0e4e1b789af6
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2021-06-23 09:53:24
# Regressor Bug: 1706374
# File Overlap Count: 1
# Description:
#   Bug 1706374 - Part 11: Add NodeController component bridging IPC and Ports, r=handyman
#   
#   The NodeController and NodeChannel types act as the backbone connecting the
#   existing IPC logic and driving the ports routing code. Individual NodeChannel
#   objects wrap and respond to messages from IPC::Channel, and the NodeController
# ==============================================================================

diff -r 8bdac5eedf78 -r 7b7819250f79 ipc/glue/MessageLink.h
--- a/ipc/glue/MessageLink.h	Tue Jun 22 18:17:21 2021 +0000
+++ b/ipc/glue/MessageLink.h	Tue Jun 22 18:17:22 2021 +0000
@@ -10,9 +10,12 @@
 
 #include <cstdint>
 #include "base/message_loop.h"
+#include "mojo/core/ports/node.h"
+#include "mojo/core/ports/port_ref.h"
 #include "mozilla/Assertions.h"
 #include "mozilla/UniquePtr.h"
 #include "mozilla/ipc/Transport.h"
+#include "mozilla/ipc/ScopedPort.h"
 
 namespace IPC {
 class Message;
@@ -22,6 +25,7 @@
 namespace ipc {
 
 class MessageChannel;
+class NodeController;
 
 struct HasResultCodes {
   enum Result {
@@ -123,6 +127,38 @@
   MessageChannel* mTargetChan;
 };
 
+class PortLink final : public MessageLink {
+  using PortRef = mojo::core::ports::PortRef;
+  using PortStatus = mojo::core::ports::PortStatus;
+  using UserMessage = mojo::core::ports::UserMessage;
+  using UserMessageEvent = mojo::core::ports::UserMessageEvent;
+
+ public:
+  PortLink(MessageChannel* aChan, ScopedPort aPort);
+  virtual ~PortLink();
+
+  void SendMessage(UniquePtr<Message> aMessage) override;
+  void SendClose() override;
+
+  bool Unsound_IsClosed() const override;
+  uint32_t Unsound_NumQueuedMessages() const override;
+
+ private:
+  class PortObserverThunk;
+  friend class PortObserverThunk;
+
+  void OnPortStatusChanged();
+
+  // Called either when an error is detected on the port from the port observer,
+  // or when `SendClose()` is called.
+  void Clear();
+
+  const RefPtr<NodeController> mNode;
+  const PortRef mPort;
+
+  RefPtr<PortObserverThunk> mObserver;
+};
+
 }  // namespace ipc
 }  // namespace mozilla
 